qiskit/transpiler/passes/routing/sabre_swap.py

```diff
@@ -329,7 +333,9 @@ def _obtain_swaps(self, front_layer, current_layout):
     def _compute_cost(self, layer, layout):
         cost = 0
         for node in layer:
-            cost += self.coupling_map.distance(layout[node.qargs[0]], layout[node.qargs[1]])
+            cost += self.coupling_map._dist_matrix[
+                layout._v2p[node.qargs[0]], layout._v2p[node.qargs[1]]
+            ]
```

Comment:
I doubt we need to get this far into the weeds, but if we're worried about the speed of access, then the `self.coupling_map` access itself isn't slotted, and we could have cached `v2p = layout._v2p` outside the loop and not really cared about the lookup penalties anyway (see main review comment as well).

Reply:
This is a good idea, I'll do some a/b testing locally to see if this makes a noticeable difference. It might, this method gets called a lot to compute the cost across multiple trials so even tiny things to reduce the overhead inside can have a surprising impact

Reply:
I made the change locally and asv shows a small improvement but not one it could distinguish from noise:

```
          391±2ms        395±0.3ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'sabre', 'noise_adaptive')
          305±2ms        308±0.7ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'sabre', 'dense')
          497±7ms        500±0.6ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'stochastic', 'sabre')
          361±4ms        363±0.6ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'sabre', 'dense')
          289±3ms          291±1ms     1.01  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'sabre', 'noise_adaptive')
          297±4ms        299±0.3ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'sabre', 'dense')
          466±4ms          468±1ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'sabre', 'noise_adaptive')
       1.16±0.01s          1.17±0s     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'sabre', 'dense')
          275±2ms        276±0.5ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'sabre', 'dense')
      4.30±0.06ms      4.31±0.06ms     1.00  mapping_passes.PassBenchmarks.time_sabre_swap(14, 1024)
          344±3ms        345±0.5ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'sabre', 'noise_adaptive')
          419±3ms        420±0.6ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'sabre', 'dense')
          681±5ms          682±2ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'sabre', 'noise_adaptive')
          255±2ms        255±0.9ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'sabre', 'dense')
          674±5ms          676±2ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'sabre', 'dense')
          297±4ms          298±1ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'sabre', 'sabre')
          393±4ms          394±1ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'stochastic', 'sabre')
          235±2ms        235±0.9ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'sabre', 'noise_adaptive')
          353±3ms          354±2ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'sabre', 'dense')
          277±1ms        277±0.6ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'sabre', 'noise_adaptive')
          339±2ms          339±1ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'sabre', 'noise_adaptive')
       1.52±0.02s          1.52±0s     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'sabre', 'dense')
       1.52±0.01s          1.51±0s     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'sabre', 'noise_adaptive')
          471±4ms          470±1ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'sabre', 'dense')
       2.37±0.01s       2.36±0.01s     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'stochastic', 'sabre')
          313±4ms          312±3ms     1.00  mapping_passes.PassBenchmarks.time_sabre_swap(5, 1024)
          446±1ms        444±0.4ms     1.00  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'sabre', 'sabre')
       1.71±0.02s          1.70±0s     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(3, 'sabre', 'sabre')
          407±4ms          404±1ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(0, 'sabre', 'sabre')
          320±2ms          318±1ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(0, 'stochastic', 'sabre')
          816±7ms          811±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'stochastic', 'sabre')
          2.30±0s          2.28±0s     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'stochastic', 'sabre')
          703±7ms          698±5ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(2, 'sabre', 'sabre')
          703±2ms          698±4ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'stochastic', 'sabre')
          681±4ms          676±1ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'sabre', 'sabre')
          669±5ms          664±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'stochastic', 'sabre')
          816±7ms          809±3ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'sabre', 'sabre')
       1.07±0.01s          1.07±0s     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'sabre', 'noise_adaptive')
         88.9±1ms       88.1±0.5ms     0.99  queko.QUEKOTranspilerBench.time_transpile_bigd(2, 'sabre')
          463±2ms          458±1ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'stochastic', 'sabre')
        339±0.9ms          336±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'sabre', 'dense')
          619±4ms          613±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(1, 'sabre', 'sabre')
          504±2ms          498±3ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(0, 'stochastic', 'sabre')
          373±3ms          369±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'sabre', 'sabre')
          266±2ms          263±1ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(1, 'sabre', 'noise_adaptive')
          568±5ms          562±2ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_180(1, 'sabre', 'sabre')
          324±4ms          320±1ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'sabre', 'noise_adaptive')
          802±4ms          792±7ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(2, 'stochastic', 'sabre')
       2.09±0.04s       2.06±0.01s     0.99  mapping_passes.PassBenchmarks.time_sabre_layout(5, 1024)
         905±10ms          894±3ms     0.99  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(3, 'stochastic', 'sabre')
          312±3ms          308±5ms     0.99  queko.QUEKOTranspilerBench.time_transpile_bigd(3, 'sabre')
      29.5±0.08ms      29.1±0.03ms     0.99  queko.QUEKOTranspilerBench.time_transpile_bigd(0, 'sabre')
       70.8±0.8ms       69.7±0.4ms     0.98  queko.QUEKOTranspilerBench.time_transpile_bigd(1, 'sabre')
       4.43±0.06s       4.36±0.04s     0.98  mapping_passes.PassBenchmarks.time_sabre_swap(20, 1024)
          436±2ms          428±2ms     0.98  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_cnt3_5_179(2, 'sabre', 'sabre')
       5.23±0.02s       5.13±0.04s     0.98  queko.QUEKOTranspilerBench.time_transpile_bntf(3, 'sabre')
       1.66±0.01s          1.62±0s     0.98  transpiler_qualitative.TranspilerQualitativeBench.time_transpile_time_qft_16(3, 'sabre', 'sabre')
       8.11±0.01s       7.93±0.01s     0.98  queko.QUEKOTranspilerBench.time_transpile_bss(3, 'sabre')
       2.28±0.03s       2.21±0.01s     0.97  queko.QUEKOTranspilerBench.time_transpile_bss(0, 'sabre')
       3.38±0.03s       3.27±0.01s     0.97  queko.QUEKOTranspilerBench.time_transpile_bss(1, 'sabre')
       3.60±0.04s       3.48±0.01s     0.97  queko.QUEKOTranspilerBench.time_transpile_bss(2, 'sabre')
       7.92±0.07s       7.66±0.01s     0.97  mapping_passes.PassBenchmarks.time_sabre_layout(14, 1024)
       2.23±0.02s       2.15±0.01s     0.96  queko.QUEKOTranspilerBench.time_transpile_bntf(2, 'sabre')
       1.26±0.01s       1.21±0.01s     0.96  queko.QUEKOTranspilerBench.time_transpile_bntf(0, 'sabre')
       2.07±0.01s       1.98±0.01s     0.96  queko.QUEKOTranspilerBench.time_transpile_bntf(1, 'sabre')
       30.3±0.04s        28.9±0.1s     0.95  mapping_passes.PassBenchmarks.time_sabre_layout(20, 1024)
```

I'm going to make the change anyway though because I think it's cleaner looking

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7036/commits/0fc21a25b5f878899f46f54386b85bfdf1256549

Reply:
Yeah, I'm not super surprised it's indistinguishable - I wouldn't be surprised if the Python interpreter does some caching of its own in loops like that.